pkg/operator/ceph/cluster/crash/reconcile.go

```diff
@@ -159,3 +167,42 @@ func (r *ReconcileNode) cephPodList() ([]corev1.Pod, error) {
 
 	return cephPods, nil
 }
+
+func (r *ReconcileNode) getCephVersion(cephCluster cephv1.CephCluster) (*version.CephVersion, error) {
```

Comment:
This should be unnecessary. The Ceph version on the label should be the Ceph version of the image being used, the one detected at operator start time.

Reply:
This might be less straightforward than I was hoping given how `register_controllers.go` is written, but in theory this should look something like starting the object or file controller. `FilesystemController` for example has the Ceph version from `clusterInfo`. @leseb wrote this code, so he might have a better idea of how to properly patch that in.

Reply:
Right; I felt like there should be a saner way to do this, but since the reconcile loop iterates over many clusters (conceivably at different versions) we do need to scrape it from somewhere. If there is an object I can spin up, that'd be great, but given that the deployment is created inside the reconcile loop this was the least baroque way I saw to implement with what I had in scope. (The other I saw was to find a pod with an image matching the crash-collector's, then map that back to its deployment, but that seemed more cumbersome.)

If there is a cleaner way to implement, I'd be quite glad of it; I agree that this is a lot (and I do have concerns about the viability of this approach during an upgrade process).

Reply:
The ceph version should always be the version of Ceph which is running in the pod's container. I think the most reliable and correct thing to do would be to run the job to detect the ceph version of the image before updating the crashcontroller.

Reply:
Okay; I thought of doing that, but I saw two problems.

First, since the deployment is created in the reconcile loop, it may not yet have been created, In this case we'd have to either spawn a new detect-version pod just for this deployment or have special logic to leave the version blank until a crash-collector pod had spawned under the deployment.

Second, it seemed a bit heavy from a resource standpoint. Still, I was considering it as though spawning a new detect-version pod would be needed (potentially for every deployment on every reconcile pass). If we can run the job without spawning the pod, that's a lot saner.

Thoughts?

Reply:
Sorry for the late reply.

> Right; I felt like there should be a saner way to do this, but since the reconcile loop iterates over many clusters (conceivably at different versions) we do need to scrape it from somewhere.

Where do you see that? The crash-collector is a namespaced resource and Rook only supports a single cluster per namespace. We even have this check here: https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/crash/reconcile.go#L115-L118 if something is weird.

> If there is an object I can spin up with a proper version.CephVersion, that'd be great, but given that the deployment is created inside the reconcile loop this was the least baroque way I saw to implement with what I had in scope. (The other I saw was to find a pod with an image matching the crash-collector's, then map that back to its deployment, but that seemed more cumbersome.)
> 
> If there is a cleaner way to implement, I'd be quite glad of it; I agree that this is a lot (and I do have concerns about the viability of this approach during an upgrade process).

Can we try to register the ceph version based on the label when we find Ceph pod around here: https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/crash/reconcile.go#L122

So we can do something like:

```golang
		for _, cephPod := range cephPods {
			if cephPod.Spec.NodeName == request.Name {
				logger.Debugf("cephPod.Spec.NodeName is %q and request.Name is %q", cephPod.Spec.NodeName, request.Name)
				hasCephPods = true
				for _, podToleration := range cephPod.Spec.Tolerations {
					// Add toleration to the map
					uniqueTolerations.Add(podToleration)
				}
				cv := cephPod.Labels["ceph-version"]
			}
		}
```

And then you can pass `cv` to `createOrUpdateCephCrash()`.

How does that sound?

Reply:
Or maybe better, you can add a new filed to the reconcile struct for the ceph version. Wether you prefer.

Reply:
Overall I see the design being that we should have a way for controllers to store state that another controller can consume. In this case, the CephCluster controller is going to detect the version of the image and store it somewhere in memory such that the crash controller can read that version. The main thing to be careful of is that this is done safely across goroutines. 

@leseb What is the upgrade path for the crash pods? Does the crash controller also watch for changes to the CephCluster CR so it knows when to update them with the new image? 

How about the following sequence:
-  The CephCluster is created with Ceph version `14.2.4`. Mons, OSDs, and other daemons are created.
- The CephCluster controller saves `cephVersions[image] = 14.2.4` in a package variable
-  The crash controller sees that daemons were created on some nodes and starts the crash pods on those nodes. The version label is added to the deployments since the controller can look up the value in `cephVersions[image]`. 

Next, let's say the CephCluster CR is updated to `14.2.5`:
- The CephCluster controller kicks off an orchestration. The new version is detected and ceph daemons are restarted.
- The crash controller notices the updated CephCluster image. The controller waits until `cephVersions[image]` is populated with the detected version for the new image, then proceeds to update the crash deployments with the new labels and image.